x-pack/plugins/fleet/server/routes/agent_policy/handlers.ts

```diff
@@ -109,37 +123,55 @@ export const createAgentPolicyHandler: FleetRequestHandler<
   const esClient = context.core.elasticsearch.client.asInternalUser;
   const user = (await appContextService.getSecurity()?.authc.getCurrentUser(request)) || undefined;
   const withSysMonitoring = request.query.sys_monitoring ?? false;
+  const { has_fleet_server: hasFleetServer, ...newPolicy } = request.body;
   const spaceId = context.fleet.spaceId;
   try {
-    // eslint-disable-next-line prefer-const
-    let [agentPolicy, newSysPackagePolicy] = await Promise.all([
-      agentPolicyService.create(soClient, esClient, request.body, {
+    if (hasFleetServer) {
+      // install fleet server package if not yet installed
+      await ensureInstalledPackage({
+        savedObjectsClient: soClient,
+        pkgName: FLEET_SERVER_PACKAGE,
+        esClient,
+      });
+    }
+    if (withSysMonitoring) {
+      // install system package if not yet installed
+      await ensureInstalledPackage({
+        savedObjectsClient: soClient,
+        pkgName: FLEET_SYSTEM_PACKAGE,
+        esClient,
+      });
+    }
+    if (request.body.monitoring_enabled?.length) {
+      // install elastic agent package if not yet installed
+      ensureInstalledPackage({
+        savedObjectsClient: soClient,
+        pkgName: FLEET_ELASTIC_AGENT_PACKAGE,
+        esClient,
+      });
+    }
+    const agentPolicy = await agentPolicyService.create(soClient, esClient, newPolicy, {
+      user,
+    });
+
+    // Create the fleet server package policy and add it to agent policy.
+    if (hasFleetServer) {
+      await createPackagePolicy(soClient, esClient, agentPolicy, FLEET_SERVER_PACKAGE, {
+        spaceId,
         user,
-      }),
-      // If needed, retrieve System package information and build a new package policy for the system package
-      // NOTE: we ignore failures in attempting to create package policy, since agent policy might have been created
-      // successfully
-      withSysMonitoring
-        ? packagePolicyService
-            .buildPackagePolicyFromPackage(soClient, FLEET_SYSTEM_PACKAGE)
-            .catch(() => undefined)
-        : undefined,
-    ]);
+      });
+    }
 
     // Create the system monitoring package policy and add it to agent policy.
-    if (withSysMonitoring && newSysPackagePolicy !== undefined && agentPolicy !== undefined) {
-      newSysPackagePolicy.policy_id = agentPolicy.id;
-      newSysPackagePolicy.namespace = agentPolicy.namespace;
-      newSysPackagePolicy.name = await incrementPackageName(soClient, FLEET_SYSTEM_PACKAGE);
-
-      await packagePolicyService.create(soClient, esClient, newSysPackagePolicy, {
+    if (withSysMonitoring) {
+      await createPackagePolicy(soClient, esClient, agentPolicy, FLEET_SYSTEM_PACKAGE, {
         spaceId,
         user,
-        bumpRevision: false,
       });
     }
 
     await agentPolicyService.createFleetServerPolicy(soClient, agentPolicy.id);
+    ensureDefaultEnrollmentAPIKeysExists(soClient, esClient);
```

Comment:
This needs to be `await`-ed. I think we may want this to happen before the policy is deployed so that if Fleet Server is trying to enroll, its enrollment key exists.

Reply:
done